src/java/org/apache/cassandra/transport/Message.java

```diff
@@ -225,12 +237,32 @@ protected boolean isTraceable()
         /**
          * Returns the time elapsed since this request was created. Note that this is the total lifetime of the request
          * in the system, so we expect increasing returned values across multiple calls to elapsedTimeSinceCreation.
+         * The elaspse time is calculated using the request time from the custom payload if it exists, otherwise it falls
+         * back to the time the request was created in the NTR stage (that is, when it was deserialized).
          *
          * @param timeUnit the time unit in which to return the elapsed time
          * @return the time elapsed since this request was created
          */
         protected long elapsedTimeSinceCreation(TimeUnit timeUnit)
         {
+            Map<String, ByteBuffer> customPayload = getCustomPayload();
+            if (customPayload != null && customPayload.containsKey(REQUEST_CREATE_NANOS))
+            {
+                ByteBuffer requestCreateNanosBuffer = customPayload.get(REQUEST_CREATE_NANOS);
+                try
+                {
+                    long requestCreationEpochNanos = ByteBufferUtil.toLong(requestCreateNanosBuffer);
```

Comment:
I misspoke regarding the usage of nanos in C*. While it is generally true, it only makes sense when calculating elapsed times against the same clock.

But since we are communicating epochs times here, the original precision (epoch millis) actually makes more sense (the clock also only exposes a translation method from millis)